icu4j/main/core/src/main/java/com/ibm/icu/message2/MFParser.java

```diff
@@ -571,17 +571,18 @@ private MFDataModel.SelectMessage getMatch(List<MFDataModel.Declaration> declara
         // Look for selectors
         List<MFDataModel.Expression> expressions = new ArrayList<>();
         while (true) {
-            // Whitespace not required between selectors:
-            // match 1*([s] selector)
-            // Whitespace not required before first variant:
-            // matcher = match-statement 1*([s] variant)
-            skipOptionalWhitespaces();
-            MFDataModel.Expression expression = getPlaceholder();
-            if (expression == null) {
+            // Whitespace required between selectors but not required before first variant.
+            skipMandatoryWhitespaces();
+            int cp = input.peekChar();
+            if (cp != '$') {
                 break;
             }
-            checkCondition(
-                    !(expression instanceof MFDataModel.Markup), "Cannot do selection on markup");
+            MFDataModel.VariableRef variableRef = getVariableRef();
+            if (variableRef == null) {
+                break;
+            }
+            MFDataModel.Expression expression =
+                    new MFDataModel.VariableExpression(variableRef, null, new ArrayList<>());
```

Comment:
Is it possible to change the data model so that selectors are constrained to be `Variable`s rather than `Expression`s? (If you're trying to avoid a public API change, I understand... although that means it's possible to construct a data model that doesn't correspond to any syntactically correct message.)